Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/WMB.hs

```diff
@@ -72,4 +332,67 @@ postUiWmbTripEnd ::
     API.Types.UI.WMB.TripEndReq ->
     Environment.Flow API.Types.UI.WMB.TripEndResp
   )
-postUiWmbTripEnd = do error "Logic yet to be decided"
+postUiWmbTripEnd (person, _, merch_city_id) tripTransactionId obj = do
+  driverId <- fromMaybeM (InternalError "Driver ID not found") person
+  fleetDriverAssociation <- FDA.findByDriver driverId >>= fromMaybeM (InternalError "Fleet Association not found")
+  transaction <- QTT.findByTransactionId (Kernel.Types.Id.Id tripTransactionId) >>= fromMaybeM (InternalError "no trip transaction found")
+  routeInfo <- QRM.findByRouteCode transaction.routeCode >>= fromMaybeM (InternalError "Route not found")
+  transporterConfig <- QTC.findByMerchantOpCityId merch_city_id >>= fromMaybeM (InternalError "Config not found")
+  fleetConfigInfo <- QFC.findByPrimaryKey (Kernel.Types.Id.Id fleetDriverAssociation.fleetOwnerId) >>= fromMaybeM (InternalError "Fleet Config Info not found")
+  now <- getCurrentTime
+
+  let distanceLeft = KU.distanceBetweenInMeters obj.location routeInfo.endPoint
+      updateAndRespond response = do
```

Comment:
its better to keep these kind of helper functions in where clause for better readability